xray/plot/facetgrid.py

```diff
@@ -0,0 +1,362 @@
+                warnings.warn("Can't use col_wrap when both col and row are passed")
+        elif row and not col:
+            self._group = row
+            self._margin_titles = False
+        elif not row and col:
+            self._group = col
+            self._margin_titles = False
+        else:
+            raise ValueError('Pass a coordinate name as an argument for row or col')
+
+        # Compute grid shape
+        if self._group:
+            self.nfacet = len(darray[self._group])
+            if col:
+                # TODO - could add heuristic for nice shape like 3x4
+                self._ncol = self.nfacet
+            if row:
+                self._ncol = 1
+            if col_wrap is not None:
+                # Overrides previous settings
+                self._ncol = col_wrap
+            self._nrow = int(np.ceil(self.nfacet / self._ncol))
+
+        self.fig, self.axes = plt.subplots(self._nrow, self._ncol,
+                sharex=True, sharey=True)
+
+        # Set up the lists of names for the row and column facet variables
+        if row is None:
+            row_names = []
+        else:
+            row_names = list(darray[row].values)
+
+        if col is None:
+            col_names = []
+        else:
+            col_names = list(darray[col].values)
+
+        self.row_names = row_names
+        self.col_names = col_names
+
+        # Next the private variables
+        self._row_var = row
+        self._col_var = col
+        self._col_wrap = col_wrap
+
+        self.set_titles()
+
+    def __iter__(self):
+        return self.axes.flat
+
+    def map_dataarray(self, func, *args, **kwargs):
+        """Apply a plotting function to each facet's subset of the data.
+
+        Differs from Seaborn style - requires the func to know how to plot a
+        dataarray.
+        
+        For now I'm going to write this assuming func is an xray 2d
+        plotting function
+
+        Parameters
+        ----------
+        func : callable
+            A plotting function with the first argument an xray dataarray 
+        args :
+            positional arguments to func
+        kwargs :
+            keyword arguments to func
+
+        Returns
+        -------
+        self : object
+            Returns self.
+
+        """
+        import matplotlib.pyplot as plt
+
+        defaults = dict(add_colorbar=False,
+                add_labels=False,
+                vmin=float(self.darray.min()),
+                vmax=float(self.darray.max()),
+                )
+
+        defaults.update(kwargs)
+
+        if self._group:
+            # TODO - bug should groupby _group
+            for ax, (name, data) in zip(self, self.darray.groupby(self.col)):
+                plt.sca(ax)
+                mappable = func(data, *args, **defaults)
+                #plt.title('{coord} = {val}'.format(coord=self.col,
+                #    val=str(name)[:10]))
+        else:
+            # Looping over the indices helps keep sanity
```

Comment:
Maybe you can write something like the `facet_data` method from Seaborn's FacetGrid? That seems like a nice separation of the iteration logic from the plotting logic